The error on Windows `CSDiagnostics.cpp(3432): error C3535: cannot deduce type for 'auto *' from 'swift::constraints::TypedNode'` ... so I think we should put the `get<>` back?

--- a/lib/Sema/CSDiagnostics.cpp
+++ b/lib/Sema/CSDiagnostics.cpp
@@ -3236,6 +3236,9 @@ bool MissingMemberFailure::diagnoseAsError() {
 
   if (diagnoseForDynamicCallable())
     return true;
+  
+  if (diagnoseInLiteralCollectionContext())
+    return true;
 
   auto baseType = resolveType(getBaseType())->getWithoutSpecifierType();
 
@@ -3421,6 +3424,37 @@ bool MissingMemberFailure::diagnoseForDynamicCallable() const {
     return true;
   }
 
+  return false;
+}
+
+bool MissingMemberFailure::diagnoseInLiteralCollectionContext() const {
+  auto &cs = getConstraintSystem();
+  auto *expr = castToExpr<Expr>(getAnchor());
+  auto *parentExpr = findParentExpr(expr);
+  auto &solution = getSolution();
+
+  if (!(parentExpr && isa<UnresolvedMemberExpr>(expr)))
+    return false;
+
+  auto parentType = getType(parentExpr);
+
+  if (!cs.isCollectionType(parentType) && !parentType->is<TupleType>())
+    return false;
+
+  if (isa<TupleExpr>(parentExpr)) {
+    parentExpr = findParentExpr(parentExpr);
+    if (!parentExpr)
+      return false;
+  }
+
+  if (auto *defaultableVar =
+          cs.getType(parentExpr)->getAs<TypeVariableType>()) {
+    if (solution.DefaultedConstraints.count(
+            defaultableVar->getImpl().getLocator()) != 0) {
+      emitDiagnostic(diag::unresolved_member_no_inference, getName());
+      return true;
+    }
+  }
   return false;
 }
 

--- a/lib/Sema/CSDiagnostics.h
+++ b/lib/Sema/CSDiagnostics.h
@@ -1076,6 +1076,10 @@ private:
   /// overload to be present, but a class marked as `@dynamicCallable`
   /// defines only `dynamicallyCall(withArguments:)` variant.
   bool diagnoseForDynamicCallable() const;
+  
+  /// Tailored diagnostics for collection literal with unresolved member expression
+  /// that defaults the element type. e.g. _ = [.e]
+  bool diagnoseInLiteralCollectionContext() const;
 
   static DeclName findCorrectEnumCaseName(Type Ty,
                                           TypoCorrectionResults &corrections,

--- a/test/Constraints/members.swift
+++ b/test/Constraints/members.swift
@@ -664,3 +664,12 @@ func test_34770265(_ dict: [Int: Int]) {
   dict.rdar_34770265_val()
   // expected-error@-1 {{referencing instance method 'rdar_34770265_val()' on 'Dictionary' requires the types 'Int' and 'String' be equivalent}}
 }
+
+// SR-12672
+_ = [.e] // expected-error {{reference to member 'e' cannot be resolved without a contextual type}}
+let _ : [Any] = [.e] // expected-error {{type 'Any' has no member 'e'}}
+_ = [1 :.e] // expected-error {{reference to member 'e' cannot be resolved without a contextual type}}
+_ = [.e: 1] // expected-error {{reference to member 'e' cannot be resolved without a contextual type}}
+let _ : [Int: Any] = [1 : .e] // expected-error {{type 'Any' has no member 'e'}}
+let _ : (Int, Any) = (1, .e) // expected-error {{type 'Any' has no member 'e'}}
+_ = (1, .e) // expected-error {{cannot infer contextual base in reference to member 'e'}}

